src/eng_back.c

```diff
@@ -1085,7 +1094,7 @@ int ctx_engine_ctrl(ENGINE_CTX *ctx, int cmd, long i, void *p, void (*f)())
 	case CMD_FORCE_LOGIN:
 		return ctx_ctrl_force_login(ctx);
 	case CMD_RE_ENUMERATE:
-		return ctx_enumerate_slots(ctx, ctx->pkcs11_ctx);
+		return ctx_enumerate_slots(ctx, ctx->pkcs11_ctx, 1);
```

Comment:
Why not just add locking around this call? This would make this single hunk change. Or add separate helper function that does slot enumeration with lock held. Adding flag whether or not to do locking is usually error prone.

Reply:
That would work too. Go ahead and add equivalent code to your other changes. 

I am not sure if I agree with: "Adding flag whether or not to do locking is usually error prone." Not having the flag got to this error, that ctx_enumerate_slots could be called while some other  thread was using the `ctx->slot_list` to load keys or certificates.